torch/distributions/transforms.py

```diff
@@ -0,0 +1,231 @@
+        """
+        Abstract method to compute inverse transformation.
+        """
+        raise NotImplementedError
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Computes the log det jacobian `log |dy/dx|` given input and output.
+        """
+        raise NotImplementedError
+
+
+class InverseTransform(Transform):
+    """
+    Inverts a single :class:`Transform`.
+    """
+    def __init__(self, bijector):
+        self.bijector = bijector
+
+    @constraints.dependent_property
+    def domain(self):
+        return self.bijector.codomain
+
+    @constraints.dependent_property
+    def codomain(self):
+        return self.bijector.domain
+
+    def forward(self, x):
+        return self.bijector.inverse(x)
+
+    def inverse(self, y):
+        return self.bijector.forward(y)
+
+    def log_abs_det_jacobian(self, x, y):
+        return -self.bijector.log_abs_det_jacobian(y, x)
+
+
+class ExpTransform(Transform):
+    """
+    Transform via the mapping `y = exp(x)`.
+    """
+    domain = constraints.real
+    codomain = constraints.positive
+
+    def _forward(self, x):
+        return x.exp()
+
+    def _inverse(self, y):
+        return y.log()
+
+    def log_abs_det_jacobian(self, x, y):
+        return x
+
+
+class SigmoidTransform(Transform):
+    """
+    Transform via the mapping `y = sigmoid(x)` and `x = logit(y)`.
+    """
+    domain = constraints.real
+    codomain = constraints.unit_interval
+
+    def _forward(self, x):
+        return sigmoid(x)
+
+    def _inverse(self, y):
+        return y.log() - (-y).log1p()
+
+    def log_abs_det_jacobian(self, x, y):
+        return -(y.reciprocal() + (1 - y).reciprocal()).log()
+
+
+class AbsTransform(Transform):
+    """
+    Transform via the mapping `y = abs(x)`
+    """
+    domain = constraints.real
+    codomain = constraints.positive
+
+    def _forward(self, x):
+        return x.abs()
+
+
+class AffineTransform(Transform):
+    """
+    Transform via the pointwise affine mapping `y = loc + scale * x`.
+
+    Args:
+        loc (Tensor or Variable): Location parameter.
+        scale (Tensor or Variable): Scale parameter.
+        event_dim (int): Optional size of `event_shape`. This should be zero
+            for univariate random variables, 1 for distributions over vectors,
+            2 for distributions over matrices, etc.
+    """
+    domain = constraints.real
+    codomain = constraints.real
+
+    def __init__(self, loc, scale, event_dim=0):
+        super(AffineTransform, self).__init__()
+        self.loc = loc
+        self.scale = scale
+        self.event_dim = event_dim
+
+    def _forward(self, x):
+        return self.loc + self.scale * x
+
+    def _inverse(self, y):
+        return (y - self.loc) / self.scale
+
+    def log_abs_det_jacobian(self, x, y):
+        result = torch.abs(self.scale).log()
+        shape = x.shape
+        for _ in range(self.event_dim):
+            result = result.sum(-1)
+            shape = shape[:-1]

```

Comment:
This would be faster:
```python
result = torch.abs(self.scale).log()
shape = x.shape
if self.event_dim:
    # NOTE: no need for contiguous here
    result = result.view(*result.size()[:-self.event_dim], -1).sum(-1)
    shape = shape[:-self.event_dim]
return results.expand(shape)
```

Reply:
Python2 complains about `result.view(*result.size()[:-self.event_dim], -1)`